src/test/java/ch/wisv/areafiftylan/products/service/TicketServiceTest.java

```diff
@@ -0,0 +1,664 @@
+        Ticket ticket = persistTicket();
+        Long ticketId = ticket.getId();
+
+        TicketTransferToken ttt = ticketService.setupForTransfer(ticketId, goalUsername);
+
+        assertEquals(ticket, ttt.getTicket());
+        assertEquals(goalUser, ttt.getUser());
+        assertEquals(true, ttt.isValid());
+    }
+
+    @Test
+    public void setupForTransferDuplicate() {
+        User goalUser = persistUser();
+        String goalUsername = goalUser.getUsername();
+        Ticket ticket = persistTicket();
+        Long ticketId = ticket.getId();
+
+        thrown.expect(DuplicateTicketTransferTokenException.class);
+        thrown.expectMessage(" is already set up for transfer!");
+
+        ticketService.setupForTransfer(ticketId, goalUsername);
+        ticketService.setupForTransfer(ticketId, goalUsername);
+    }
+
+    @Test
+    public void setupForTransferAlreadyLinked() {
+        User goalUser = persistUser();
+        String goalUsername = goalUser.getUsername();
+        Ticket ticket = persistTicket();
+        Long ticketId = ticket.getId();
+        testEntityManager.persist(new RFIDLink("1234567890", ticket));
+
+        thrown.expect(TicketAlreadyLinkedException.class);
+        thrown.expectMessage("Ticket has already been linked to a RFID");
+
+        ticketService.setupForTransfer(ticketId, goalUsername);
+    }
+
+    @Test
+    public void transferTicket() {
+        User goalUser = persistUser();
+        String goalUsername = goalUser.getUsername();
+        Ticket ticket = persistTicket();
+        Long ticketId = ticket.getId();
+        TicketTransferToken ttt = ticketService.setupForTransfer(ticketId, goalUsername);
+        String token = ttt.getToken();
+
+        ticketService.transferTicket(token);
+
+        assertEquals(goalUser, ticket.getOwner());
+        assertFalse(ttt.isUnused());
+    }
+
+    @Test
+    public void transferTicketTokenNotFound() {
+        thrown.expect(TokenNotFoundException.class);
+        thrown.expectMessage("Could not find token ");
+
+        ticketService.transferTicket("invalid_token");
+    }
+
+    @Test
+    public void transferTicketTokenNull() {
+        thrown.expect(TokenNotFoundException.class);
+        thrown.expectMessage("Could not find token ");
+
+        ticketService.transferTicket(null);
+    }
+
+    @Test
+    public void transferTicketTokenInvalid() {
+        User goalUser = persistUser();
+        String goalUsername = goalUser.getUsername();
+        Ticket ticket = persistTicket();
+        Long ticketId = ticket.getId();
+        TicketTransferToken ttt = ticketService.setupForTransfer(ticketId, goalUsername);
+        String token = ttt.getToken();
+
+        ttt.use();
+
+        thrown.expect(InvalidTokenException.class);
+        thrown.expectMessage("Token is expired, has been already used or has been revoked.");
+
+        ticketService.transferTicket(token);
+    }
+
+    @Test
+    public void transferTicketTicketAlreadyLinked() {
```

Comment:
This is a good test though, a user should not be able to transfer the ticket after the event has started when the ticket was set up for transfer before checking in.